From 74545046a1b392be03e484e1737b1786b8d2f6e0 Mon Sep 17 00:00:00 2001
From: minaripenguin <minaripenguin@users.noreply.github.com>
Date: Sat, 18 May 2024 08:01:14 +0800
Subject: [PATCH 13/21] GameManagerService: Prevent multiple threads from
 accessing settings observer onChange

* this could lead to potential increase of cpu usage in system_server

Change-Id: I5b7b606a4f8795cd59ddcb15128add8bd70b5ecb
Signed-off-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../server/app/GameManagerService.java        | 40 ++++++++++---------
 1 file changed, 21 insertions(+), 19 deletions(-)

diff --git a/services/core/java/com/android/server/app/GameManagerService.java b/services/core/java/com/android/server/app/GameManagerService.java
index 87a49a957..354151f10 100644
--- a/services/core/java/com/android/server/app/GameManagerService.java
+++ b/services/core/java/com/android/server/app/GameManagerService.java
@@ -2386,25 +2386,27 @@ public final class GameManagerService extends IGameManagerService.Stub {
 
         @Override
         public void onChange(boolean selfChange, Uri uri) {
-            String newValue = Settings.Secure.getString(mContentResolver,
-                    Settings.Secure.GAME_OVERLAY);
-            // We write key and value of the device_config property as a single string
-            // from our GameSpace.
-            // ';;' is the separator betweeen key and value.
-            // Example: com.libremobileos.game;;mode=2,downscaleFactor=0.7:mode=3,downscaleFactor=0.8
-            // So split the key and value from the string
-            // and set the device_config propery.
-            String[] parsedValues = newValue.split(";;");
-            // Value should contain both package name and config.
-            // Otherwise don't do anything.
-            if (parsedValues.length < 2) return;
-            // We don't need to care about any format and all.
-            // It will be handled by the GamePackageConfiguration while
-            // parsing the device_config property.
-            String packageName = parsedValues[0];
-            String configValue = parsedValues[1];
-            DeviceConfig.setProperty(DeviceConfig.NAMESPACE_GAME_OVERLAY,
-                    packageName, configValue, false);
+            synchronized (mDeviceConfigLock) {
+                String newValue = Settings.Secure.getString(mContentResolver,
+                        Settings.Secure.GAME_OVERLAY);
+                // We write key and value of the device_config property as a single string
+                // from our GameSpace.
+                // ';;' is the separator betweeen key and value.
+                // Example: com.libremobileos.game;;mode=2,downscaleFactor=0.7:mode=3,downscaleFactor=0.8
+                // So split the key and value from the string
+                // and set the device_config propery.
+                String[] parsedValues = newValue.split(";;");
+                // Value should contain both package name and config.
+                // Otherwise don't do anything.
+                if (parsedValues.length < 2) return;
+                // We don't need to care about any format and all.
+                // It will be handled by the GamePackageConfiguration while
+                // parsing the device_config property.
+                String packageName = parsedValues[0];
+                String configValue = parsedValues[1];
+                DeviceConfig.setProperty(DeviceConfig.NAMESPACE_GAME_OVERLAY,
+                        packageName, configValue, false);
+            }
         }
     }
 
-- 
2.53.0

